### ISSUE TITLE
Store state not showing in Input element

### ISSUE THREAD

**Issue by mmkari** [erikras/redux-form#1917]
The form detailed below contains one Field for a first name and the value of that field's input is initialized to the value of 'initialName' as specified in the `initialValues` object given to `mapStateToProps`. Upon loading the component, the store contains the given initial value. However, the input element never displays that value. How does one make the controlled component display the stored value? 

Removing the `{...input}` from the `<input>` element in `renderField` does not seem to help as then the input field is non-responsive and the form only dispatches one letter CHANGE actions, overwriting the value in the store with one letter strings. In addition, in this case the ending BLUR action overwrites the stored value with an empty string when the field is deactivated.

Another issue is that the `handleSubmit` function provided by `props` should pass the validated form values to the `onSubmit` function but instead passes the initial value of 'initialName'. Also, when there are errors, the `handleSubmit` function does not prevent passing control to `onSubmit` and no error messages are displayed in the rendered form element.

Could someone please explain why and how to fix it?
Thanks!

---

``` JavaScript
import React from 'react';
import { connect } from 'react-redux';
import { Field, reduxForm } from 'redux-form';


const renderField = ({ input, label, value, type, meta: { touched, error }, onChange }) => (
  <div>
    <label>{label}</label>
    <div>
      <input {...input} value={value} type={type} onChange={onChange} />
      {touched && error && <span>{error}</span>}
    </div>
  </div>
);


let SettingsForm = (props) => {

  const onSubmit = (formProps) => {
    alert("Name passed to onSubmit: '"+ formProps.firstName +"'");
    console.log("Form props passed by handleSubmit: ", formProps);
  }

  const { handleSubmit } = props;

  return (
      <form onSubmit={handleSubmit(onSubmit.bind(this))}>
        <div>
          <h3>Personal Info</h3>
          <div>
            <Field
              name="firstName"
              component={renderField}
              type="text"
              label="First name"
              onChange={() => {}}
              />
          </div>
        </div>
        <button type="submit">Submit</button>
      </form>
  );
}


const validate = (values) => {
  const errors = {};
  console.log("values at validation: ", values);
  if (!values.firstName) {
    errors.firstName = 'Enter your first name';
  }
  console.log("The current set of errors is", errors);
  return errors;
}


const mapStateToProps = (state) => {
  return {
    initialValues: {
      firstName: "initialName",
    },
  };
}

// redux-form decorator
SettingsForm = reduxForm({
  form: 'userSettingsForm',
  validate,
})(SettingsForm);

// Important: connect only after calling reduxForm!!!
export default connect(mapStateToProps)(SettingsForm);
```

**Comment by JosephEarl** [comment]
You've overridden the `onChange` provided by redux-form which will prevent it dispatching the correct actions as you enter text.
Try removing `onChange={() => {}}` from `Field`.

**Comment by mmkari** [comment]
Thanks for the comment @JosephEarl !

However, I've purposely overwritten the `onChange` handler so that the store would not receive so many unnecessary CHANGE actions. The store fields will get updated by the BLUR action in any case. Removing the `onChange={() => {}}` overwrite nor any other solution that I've tried solves the following problems that still persist:

1. the initial value of 'initialName' does not show up in the field even though it shows up in the store and the field is controlled by redux-form.
2. even as the value in the field is manually entered and the corresponding redux-store value `userSettingsForm > values > firstName` reflects that change correctly, upon submit the redux-form provided function `handleSubmit` neither performs validation nor passes the actual form input value (again, correctly set in store) to the passed and bound `onSubmit` function. The value that gets passed to `onSubmit` is for some inexplicable reason the initial value of the field.

It's also worth mentioning that the official examples don't work either and show the same issues relating to initialized fields not showing their values and those of the submit handler.

Has no one else encountered these issues?

**Comment by JosephEarl** [comment]
I put your code into a codepen with the `onChange` removed and it works fine for me - http://codepen.io/anon/pen/bwKzVq.

Have you added the Redux Form reducer?

Here is the same with custom `onChange` - http://codepen.io/anon/pen/VKdRwP - the field does not update when you type because no change event is dispatched.

**Comment by mmkari** [comment]
Yes, I've added the reducer in the correct `combineReducers()` call (as shown below) and the redux store's state is even changed correctly under `userSettingsForm > values`. 

`import { reducer as formReducers } from 'redux-form';`

```
export const someReducers = combineReducers({
    ...,
    form: formReducers,
});
```

The fact that the code works on _codepen_ suggests that the issue is related to some incompatibilities between redux-form and the other libraries currently in use in my setup and it seems to be an information flow issue from the store state to the respective fields. Even the `onSubmit` function is given the initial value of the field in spite of the value having been successfully set in store. 

It's difficult to continue investigating this since there are no warnings or error messages that give any indication of what conflict there may be. Therefore I will no longer continue with redux-form at this time. Others may still comment here or close the ticket as they please.

**Comment by gocamille** [comment]
@mmkari I am experiencing the exact same issue for all examples and for your project locally. I have been having the same issues with my project for two days, so I'm a bit relieved to know it's not just me. 

Once I have the time I will post the deets about my environment.

**Comment by niksajanjic** [comment]
**PS:** I'm going to assume you're not overriding onChange, so i'm going to pretend you didn't set that on **Field** component.

It doesn't work because you are destructuring your **renderField** component like like this:

    { input, label, value, type, meta: { touched, error } }

Which is wrong, because **value** is always empty. The thing is, **value** is part of **input** object (input.value, **value** is not part of your props, you can't access it props.value, it's props.input.value). So you should either remove **value** from your input like this:

    <input {...input} type={type} />

You see, no ```value={value}``` because that is going to be inserted by ```{...input}```. That would be a proper way of doing it. But if there's any reason why you need to set **value** manually on your input, than you should destructure like this:

    { input: { value, ...input }, label, type, meta: { touched, error }, onChange }

You now have **input** object that holds everything except **value**, and **value** variable that holds just **value**. And now you can do this:

    <input {...input} value={value} type={type} />

**Comment by niksajanjic** [comment]
PS: In my personal experience I had similar issue where state had value, but it was never passed to input. That was beacuse I use immutable and I imported redux-form decorator:

    import { reduxForm } from 'redux-form';

Instead of:

    import { reduxForm } from 'redux-form/immutable';

Maybe someone would find this useful. I'm going to add custom linting rule to notify me when I use wrong import.

**Comment by wayne163** [comment]
Thanks for the commen @niksajanjic 
I had the same problem, but after I changed the redux-form import from 
`import { reduxForm } from 'redux-form` 
to 
`import { reduxForm } from 'redux-form/immutable `.
It works!

The introduction about using immutableJS with redux-form is only set in the [Immutable JS Example](http://redux-form.com/6.4.3/examples/immutable/) . Maybe it's better to give a warning in console when using immutable js but with wrong reduxForm import.

**Comment by piuccio** [comment]
thanks for the last two comments. I think the example with ImmutableJS should be made more prominent especially considering that react-boilerplate uses immutable data. Took me some googling to get here

**Comment by fmigot** [comment]
Hello,
I had the same issue with the import from 'redux-form and I did the modification as @niksajanjic mentioned and now it works!
Thank you so much!

**Comment by i6mi6** [comment]
For those who don't use immutablejs don't forget to pass {...restInput} to your TextInput inside your `<Field component={this.renderInput} />` like so: 
`<TextInput onChangeText={onChange} {...restInput}` where your function looks like this
`renderInput({ input: { onChange, ...restInput }, meta: { touched, error, warning } })`

**Comment by danielrob** [comment]
I think I've seen this documentation issue on other issues as well. Would someone be willing to raise a documentation PR for this issue with immutable? Was this the complete resolution of this present github issue?

**Comment by rodolfoag** [comment]
@niksajanjic Thank you!!!

**Comment by dkniffin** [comment]
I made the same mistake as others. @niksajanjic's answer fixed half of it, but I also wanted to say: don't forget to also replace:

`import { Field } from 'redux-form';`

with

`import { Field } from 'redux-form/immutable';`

**Comment by gustavohenke** [comment]
Closing since is a mistake by most users, but feel free to open a PR with a fix if you're up for it.
